### ISSUE TITLE
Shift key randomly gets stuck down

### ISSUE THREAD

**Issue by the-wes** [symless/synergy-core#4493]
Confer https://github.com/synergy/synergy/issues/9, randomly I'll be typing along and my letters will come out as caps, without caps lock being enabled. Even pressing the shift keys on the server keyboard doesn't help, I have to reach over and press the right-hand shift key on the client keyboard to unstick. Pressing the left one does not help.

Of course, moving to the server or another client does not cause the shift key to be stuck in this case.

The shift key is never stuck down while using the client's local keyboard.

**Comment by shawend** [comment]
I have a similar problem. I have a Dell Inspiron Duo running Ubuntu 12.10 as a client and a Dell Inspiron One running Windows 7 Home Premium as the server. When the client connects I move the cursor to the client and am able to login. After moving the cursor off of the client screen to the server screen and then back to the client screen the caps lock is stuck on for the client even though  it is not pressed on the keyboard. Using either of the shift keys on the client had no effect. The left shift key on the server also didn’t have any impact. However, once you use the right shift key on the server everything is magically fixed. Both keyboards work correctly and moving the cursor from machine to machine does not bring the incorrect keyboard function back. Once the client is re-booted then the problem comes back again but it is cleared up with the right shift key.

**Comment by astehlik** [comment]
I can confirm this problem. My environment looks like this:

Server is a Kubuntu 64bit system (problem occurs with 14.04, 14.10 and 15.04).

Client is Windows 7 32bit.

The problem only occurs on the client. The <kbd>Shift</kbd> key seems to be pressed permanently on the client (randomly). Sometimes additionally the <kbd>Strg</kbd> key is stuck. Symtoms are:

* When clicking in a list (file list in windows explorer, multiselect box), multiple items get selected
* When clicking on a program in the taskbar, a new instance is opened, when the <kbd>Strg</kbd> Key is also stuck the new instance will be run as Administrator.

Pressing the **right** <kbd>Strg</kbd> / <kbd>Shift</kbd> keys on the keyboard of the client makes them unstuck.

**Comment by beporter** [comment]
See also these related issues:

* Left mouse button gets stuck in right click mode on OS X client #3823
* Ctrl or shift keys randomly simulated while in IE #3194
* Command key stuck (Mac client, Windows server) #4057
* Arrow keys and ctrl+arrow causes screen rotation on Windows #4449
* Modifier state becomes out of sync on screen lock #4519

A fix was just merged specifically for Mac clients (Modifier keys often stuck down on Mac client #4564), but this problem is cross-platform with possible-different root causes on each and probably still needs more attention.

**Comment by astehlik** [comment]
Thank you for the hint!

I think I'm facing the exact same issue that is described in #4519. I also use a combination <kbd>Strg</kbd> + <kbd>Shift</kbd> + <kbd>L</kbd> to lock my Windows Screen and after I login both modifier keys are stuck.

**Comment by XinyuHou** [comment]
> Server: Synergy 1.7.6 on Windows 7 (64-bit)
Client: Synergy 1.7.6 on Mac OS El Capitan (10.11.4)

Since upgrading the server from 1.7.5 to 1.7.6 whenever the client connects, the keybinding I set on my Logitech mouse (a thumb-button set to "CTRL-W"; set through SetPoint Control Center 6.67.83) stops working correctly. (Other bindings, such as doubleclick, work fine; double-checked that it's not the button itself.)
The command "CTRL-W" doesn't work at all—sometimes, though, the CTRL-key gets stuck (with all the obnoxious side-effects that creates).

Current workaround: Fire up Synergy on the server, stop it, then restart it. Everything works fine afterwards.
The error arises again when disconnecting and then reconnecting the client.

Issue was not present with 1.7.5 as a server.

(Currently figuring out minimal reproduction steps and how to generate relevant log-files.)

**Comment by sim590** [comment]
I still experience the issue on 1.8.8 (both client and server). Has there been a fix merged at some point on a subsequent version?

**Comment by aboudreault** [comment]
So what's the plan about this issue?

**Comment by nbolton** [comment]
macOS specific: https://github.com/symless/synergy-core/issues/4493

**Comment by Smurf** [comment]
With newer Linux kernels (5.4/5.5) this is worse than ever.

Open source forks of Synergy have fixed this issue. Feels a bit like the money spent on a license was useless if bugs that are 5+ years old are still around.

**Comment by nbolton** [comment]
@Smurf Hi, Synergy is still open source. Could you link to the issue/commit that fixes this in other forks? We can easily merge it in.

**Comment by panchopin** [comment]
I just bought this software 2 days ago and just found this issue because it literally happens all the time. Impossible to work with! will ask for a refund!

**Comment by laino** [comment]
Yeah this stuff happens all the time. I can't remember 10 minutes passing without some modifier (Super/Shift/Alt/whatever) getting stuck.

**Comment by jkfran** [comment]
I have this problem as well. MacOS as host and Ubuntu 22.04. The modifier keys are stuck pressed constantly :S

**Comment by nbolton** [comment]
> Impossible to work with! will ask for a refund!

Sorry to hear that. Did you get your refund ok? Once we fix it, I hope you can come back to us.

**Comment by theLucasAntunes** [comment]
Is this still an issue? I see this is now the upstream of Synergy, and this issue is an OOOOOOOOLD issue.

I first spotted it in Barrier but discovered it is just everywhere, all packages have it.

Sad to see it's been all this time and it's not been fixed yet!

**Comment by the-wes** [comment]
I am not aware of a specific fix, however I have not seen anyone report experiencing this issue in a couple of years. It's extremely difficult to reproduce, so I am not really surprised it hasn't been fixed. It will take some significant effort to put this one out, and without it affecting a large number of people, it will be tough to justify putting that much effort into it.

**Comment by nbolton** [comment]
> I have not seen anyone report experiencing this issue in a couple of years

Hey @the-wes 👋🙂 -- thanks for the insight. I hadn't seen it for few years, but recently with Wayland--libei, not XWayland--I think the may have had a resurgence. I have experienced some keys being stuck when using my Wayland GNOME machine as server (using libei).

@sithlord48 Have you seen this at all with your Wayland server?

**Comment by sithlord48** [comment]
I think if hold a modifier when you move to another screen it can get stuck on until you go back to the host. Idk its  an os specific issue at all

**Comment by sithlord48** [comment]
I have noticed on my laptop keyboard when this does happen when i hit shift the caps lock light does a quick pulse.. maybe or maybe not be relevant

**Comment by airtonix** [comment]
still happening. super annoying

**Comment by nbolton** [comment]
> still happening. super annoying

Thanks for confirming. What OS and Deskflow version are you using?

**Comment by underclockeddev** [comment]
Happening to me.
Smashing the stuck modifier, left and right, on either the server or the host, does not resolve. Only way I've found to resolve is to end the server, restart it, then reconnect.

**Server**:               
Deskflow: 1.24.0.0 (56a1bd68)
Qt: 6.9.2
System: EndeavourOS
Session: KDE (wayland)


**Client**:
Deskflow: 1.24.0.0 (56a1bd68)
Qt: 6.9.0
System: Windows 11 Version 22H2

**Comment by sithlord48** [comment]
With the recent fixes #9062, and with shift restored in #9074  this may finally be fixed (for wayland servers)

**Comment by cephcyn** [comment]
Echoing two comments ago to say that I just observed this issue (shift key held down on client after some input from the server, no amount of pressing shift on client or server actually fixed this issue and I had to reboot the client) with:

Server:
Deskflow: 1.24.0.219 (3af83dd9)
Qt: 6.10.0
System: Windows 10 Version 22H2

Client:
Deskflow: 1.24.0.0 (56a1bd68)
Qt: 6.9.3
System: KDE Flatpak runtime
Session: ubuntu:GNOME (x11)

**Comment by crakerjac** [comment]
Came here to say I'm seeing the same thing... but it's usually with the SUPER button.  Not something I'm holding down when switching between screens.

I have a Windows 11 Pro (Server) and Linux Mint 22 (Client).  Both running DeskFlow: 1.24.0

When in this state, restarting DeskFlow on either the Server or Client fixes the issue for a minute or two, but then comes right back.  To fully resolve the issue, I need to reboot.

Does this sound like the same bug or something different?

**Comment by sithlord48** [comment]
It could this same bugs the Fixes I referred to only fix it for wayland servers (maybe clients ?)

**Comment by crakerjac** [comment]
Good call.  I'm running X11 by default.  I've updated my previous comment.

**Comment by kori4r2** [comment]
Experiencing a similar issue. Both server and client are running arch linux, KDE Wayland and using version 1.24.0. There are two separate issues that happen: The first one is the control key getting stuck as pressed on the client regardless of me actually pressing it. The other makes it so that whenever I try to press one of the F keys it changes tty as if I had pressed a ctrl+alt+F key shortcut. The weird part is that when the sticky ctrl key bug is happening it triggers other shortcuts but not the one to change tty, and when the F keys bug is happening it is not because both ctrl and alt keys are sticky, since it doesn't trigger other keyboard shortcuts that use them. I cannot figure out what triggers those bugs (could be related to system sleep but I haven't tested yet) but restarting deskflow fixes them.

**Comment by sithlord48** [comment]
@kori4r2  Please try the continuous version and report back.

**Comment by kori4r2** [comment]
@sithlord48 I saw that the new release contains the patches mentioned earlier, updated both systems and the issues haven't happened for a few days now.

**Comment by crakerjac** [comment]
+1.   I have been running 1.25.0 for a while now, and I have not seen the stuck shift key since updating.

**Comment by georgedorn** [comment]
I'm running 1.25.0 and modifier keys are still getting stuck down constantly.  It's so bad I wrote my own alias to send key up events for every modifier on the keyboard.  Most of the time, this works, but sometimes I can't even type the alias into a terminal.

I don't even need to be using that key.  Sometimes 'alt' will be stuck down when I haven't used it in minutes.

```
alias fixmods='xdotool keydown Enter && xdotool keyup Enter Shift_L Shift_R Control_L Control_R Meta_L Meta_R Alt_L Alt_R Super_L Super_R Hyper_L Hyper_R ISO_Level2_Latch ISO_Level3_Shift ISO_Level3_Latch ISO_Level3_Lock ISO_Level5_Shift ISO_Level5_Latch ISO_Level5_Lock'
```

All I need now is a way to hook into when that machine gains focus and run a script to do this automatically...

ETA:  Found one.  Maybe 1.26 will fix this, but continuous doesn't, so here's a horrible hack:
- Enable INFO-level logging in Deskflow, on the machine having the problem.
- Run this in a shell, or in a bash script or something: 
```
tail -f "$HOME/deskflow.log" | awk '/entering screen/ { system("xdotool keyup Enter Shift_L Shift_R Control_L Control_R Meta_L Meta_R Alt_L Alt_R Super_L Super_R Hyper_L Hyper_R ISO_Level2_Latch ISO_Level3_Shift ISO_Level3_Latch ISO_Level3_Lock ISO_Level5_Shift  ISO_Level5_Latch ISO_Level5_Lock && notify-send DONG") }'
```

Obviously if you aren't logging to your $HOME, change what you're tailing.

**Comment by sithlord48** [comment]
@georgedorn  what is your keyboard local and your server / client info ?

**Comment by georgedorn** [comment]
I am running macOS as the server and Manjaro as the client.  I have also run them the other way; I'm not positive, but I don't think it happens in that direction.  I don't work that way often, though.

On the mac:
Deskflow: 1.25.0.47 (13219ca4)
Qt: 6.10.0
System: macOS Sequoia (15.6.1)

On the linux desktop:
Deskflow: 1.25.0.0 (d4ff55da)
Qt: 6.10.1
System: Manjaro Linux
Session: XFCE (x11)

I have tried other versions, though.  I experienced this same issue on the other variants; it happens on Synergy and Input Leap as well.

I'm temporarily disabling my hacky script in the hopes that I can capture a log of it happening, but I cannot force it to happen, e.g. by holding modifiers while changing screens.
